### PR TITLE
subsystem: task_wdt: Implement software watchdog suitable for multiple threads

### PR DESCRIPTION
**Pull request by martinjaeger**
**Remark**: The approach described in this first comment has evolved and changed quite a bit. Consider also comments further down.

This PR adds a software watchdog that is processed within a `k_timer` callback.

Only Nordic MCUs seem to support multiple watchdog channels in hardware. However, the [watchdog API documentation](https://docs.zephyrproject.org/latest/reference/peripherals/watchdog.html#c.wdt_install_timeout) presumes that this is the case in general. At least for STM32 I was quite confused that `wdt_install_timeout` would always return a hard-coded `0` as the channel ID (instead of `-ENOMEM` after the first call).

This software watchdog implementation allows to use multiple channels even if the hardware supports only a single watchdog timeout. In this way it allows to use a watchdog properly in a multi-threaded environment. I added an implementation for the STM32 IWDG, but it could be easily added to other hardware aswell.

### Implementation details
Roughly following the approach as described [here](https://stackoverflow.com/questions/13217959/how-to-use-the-watchdog-timer-in-a-rtos/13219848#13219848).
- If the software watchdog is enabled via Kconfig, `wdt_install_timeout` forwards each call to the software watchdog implementation and returns a value >= 1 until the maximum number of channels (also set via Kconfig) is reached.
- The hardware watchdog is assigned to channel 0 and its timeout is set to the minimum timeout of all software watchdog channels to provide a fall-back.
- The hardware watchdog is not fed by the application, but automatically within the software watchdog.
- The software watchdog itself is based on a timer, which checks every channel for timeout in a regular interval.
- If the timer gets stuck for some reason, the hardware watchdog automatically becomes the fallback solution and resets the device.

### Possible improvements
- Allow multiple instances of the software watchdog. This would require a `struct sw_wdt` passed with each function call that holds the different channels instead of the static array. However, I can't see any use case where two software watchdogs would be needed.
- Make minimum interval configurable via Kconfig and save some processing cycles if 1 ms resolution is not needed.

Please let me know your thoughts.

### CODE REVIEW

**Comment by r2r0** [comment]
IMHO for low power applications 1ms timer period is overkill.
It would be really nice to have an option to set (increase) minimum interval.

**Review by nslowell — COMMENTED** [review]
I think this a good starting point.  However, some things need to be considered.

There's a small disconnect between the hardware and software watchdogs b/c what if the hardware watchdog timer is shorter than the software timer?  Seems like there needs to be some checking between them to make sure their configurations keep each other happy.

With that, I personally would like it where if I'm using a software watchdog, I don't want to have to touch the hardware watchdog at all.  So maybe the solution to this and the above issue is that the software watchdog encapsulates the hardware watchdog usage (when configured to do so, read below).

Also, I'm wondering if we want to allow this to be used even when there isn't an underlying hardware watchdog.  If the software watchdog expired and there's no hardware watchdog, I don't see why it couldn't call the system reset instead.  That would require some decoupling and/or more configuration.  

I think since Zephyr often has design for static/compile-time declarations and setup, this might need some of that to complement the run-time setup.  Allow for static declarations/registrations of channels.  This would likely help the WD to start earlier (through SYS_INIT) which was an previously expressed desire as well.

I know these last two add some complexity, but they add flexibility and could potentially come later.

**Comment by henrikbrixandersen** [comment]
@martinjaeger Have you considered how this will work on SoCs/boards with multiple HW watchdogs?
I think the binding between the hardware and the software watchdog should be left up to the application - it should not happen automatically in the hardware watchdog driver.

**Comment by martinjaeger** [comment]
> @martinjaeger Have you considered how this will work on SoCs/boards with multiple HW watchdogs?
> I think the binding between the hardware and the software watchdog should be left up to the application - it should not happen automatically in the hardware watchdog driver.

No, didn't really consider setups with multiple HW watchdogs yet. And I see this could be an issue with current approach.

Also agree that the binding between SW and HW watchdog should ideally be done in the application. However, I don't have a good idea at the moment how we could re-use existing API for SW and HW watchdog and still interconnect two of them. It would be nice to have a way to enable the SW watchdog in Kconfig and assign a specific HW watchdog to use as fallback. Afterwards the application should be able to use the watchdog API exactly as before, just that we would have more channels e.g. for STM32.

Do you have suggestions how that could be achieved in a clean way, @henrikbrixandersen?

**Comment by henrikbrixandersen** [comment]
> Do you have suggestions how that could be achieved in a clean way, @henrikbrixandersen?

Perhaps using the same API for the HW and SW watchdog implementations is not the best approach? How about instead encapsulating the SW watchdog functionality in a `wdt_context.h` (or similar named) header file under `drivers/watchdog` and then let each applicable watchdog HW driver include and enable the functionality (depending on the global Kconfig) transparent to the user - effectively increasing the number of channels for the given watchdog driver?  An approach similar to the other `*_context.h` header files currently in Zephyr.

File: drivers/watchdog/Kconfig.software
Comment (pabigot) [inline drivers/watchdog/Kconfig.software:25]: extra `for`

File: drivers/watchdog/wdt_software.h
Comment (pabigot) [inline drivers/watchdog/wdt_software.h:18]: Please document units: `timeout` is ms, `feed_time` is from `k_uptime_get()`.

**Review by pabigot — COMMENTED** [review]
The architecture of having a timer fire at a fixed interval to check whether the watchdog timeouts are all still met isn't very efficient.  It should be possible to calculate the duration until a watchdog check is required, and sleep until then.

For the overall solution I'd really rather see something that provides multiplexed software timeouts that contribute to whether a hardware watchdog is fed.  I've done this in [nrfcxx](https://pabigot.github.io/nrfcxx/classnrfcxx_1_1watchdog__extended__channel.html).  However, I'm not sure that can be generalized to Zephyr's watchdog API as it assumes a dedicated hardware channel for the extended channels.

Regarding setting the hardware timeout to the minimum of the software timeouts you should be aware that many watchdog drivers only let you configure the watchdog once.

**Comment by mbolivar-nordic** [comment]
Dev-review 8 October 2020:

Leaning towards a separate software module which feeds a hardware watchdog using a k_timer. Further discussion is needed and will take place in this PR.

**Comment by nslowell** [comment]
The idea for the software watchdog is for multisourcing kicks even though there is a single hardware watchdog.  If you've got a system with multiple hardware watchdogs, why not use them directly?  If you had more sources for kicks than hardware watchdogs, I would still think each software watchdog instance would connect with a single hardware watchdog.

There may be a need for backend configurations because I still think that a software watchdog should be able to be used when there is NO hardware watchdog at all, but I also think interacting through a software watchdog should mean you don't have to interact with the hardware watchdog (it's handled internally, perhaps through the backend).

**Comment by martinjaeger** [comment]
It took a bit longer to update this PR, but I think I addressed several shortcomings of the previous version. (please ignore the first commit, which is required for testing with STM32 and is already part of another PR)

1. The software watchdog (now called task watchdog) doesn't require any change in the hardware watchdog drivers anymore. Instead, an existing hardware watchdog can be defined as an optional fallback. The task watchdog can also be used independent of a hardware watchdog. Also moved the code to `subsys` as suggested by @nslowell.
2. The task watchdog still uses a kernel timer as its backend. But if configured properly, the timer ISR is never actually called during normal operation, as the timer is continuously updated in the `feed` calls.

To keep it simple, it's currently not possible to have multiple instances of task watchdogs. So there is only a single instance per app which has a predefined maximum number of channels.

I did some research regarding features of task-level / software watchdogs in other RTOSs. The following table compares those findings (no guarantee that everything is correct) with the proposed subsystem for Zephyr:

  | VxWorks | ESP-IDF TWDT | Segger embOS | Zephyr (RFC)
-- | -- | -- | -- | --
Context | system clock ISR | ? | SysTick_Handler suggested | kernel timer ISR
API: init |   | esp_task_wdt_init() | OS_WD_Config() | task_wdt_init(hw_dev)
API: add | wdCreate() | esp_task_wdt_add() | OS_WD_Add() | task_wdt_add(timeout)
API: feed | wdStart() | esp_task_wdt_reset() | OS_WD_Trigger() | task_wdt_feed(id)
API: cancel  | wdCancel() |   |   |  
API: delete | wdDelete() | esp_task_wdt_delete() | OS_WD_Remove() | ToDo
API: check |   | esp_task_wdt_status() |   |  
API: process |   |   | OS_WD_Check() | automatically
Default action | callback | console print | callback | callback
Callback | yes | only panic handler | yes | yes
SW Reset | ? | no | optional | optional
HW wdt fallback | ? | yes | yes | optional
Individual timeout per task |   | no | yes | yes
Timeout resolution | system clock ticks | seconds (!) | OS_TIME (system ticks) | system ticks
Reference | [link](https://comp.os.vxworks.narkive.com/tLFU3ezA/watch-dog-timer-functionality) | [link](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/wdts.html) | [link](https://www.segger.com/downloads/embos/UM01001) |  

Please let me know what you think about the new approach. I will also add further documentation for the API and the sample in the end.

**Comment by nslowell** [comment]
Looks promising.  I like it, but I know there is great focus on compile-time configuration in Zephyr, so even though the option of passing in the hardware watchdog device name into init() works, it's run-time configuration instead of compile-time, so code size will be bigger than it needs to be, especially b/c it forces wdt interface to get linked in even though it will never be used.  You might see if you can convert it do Kconfig.

Be sure to consider what will happen if init() gets called multiple times.

**Comment by mbolivar-nordic** [comment]
> option of passing in the hardware watchdog device name into init() works, it's run-time configuration instead of compile-time, so code size will be bigger than it needs to be, especially b/c it forces wdt interface to get linked in even though it will never be used. You might see if you can convert it do Kconfig.

Hmm, I think it should rather just take a `const struct device*` or something like that.

We don't want any more `CONFIG_FOO_DEV_NAME` variables in Kconfig now that we're relying on DT for things like this.

See https://docs.zephyrproject.org/latest/guides/kconfig/tips.html#what-not-to-turn-into-kconfig-options

**Comment by nslowell** [comment]
It's possible I misunderstand something, but this isn't a device.  It's a software-based watchdog subsystem that someone can decide to use or not on any platform.  The ways it's written though, it forces linking in the wdt driver interface, which shouldn't be required so that it an run on a system without a hardware watchdog.  However, since it _can_ fall back on a hardware watchdog device, perhaps it should somehow check the DT for watchdog hardware to best determine what code needs to be compiled and linked.  I'm not exactly sure how that will work b/c DT names are specific.  That's why I was assuming Kconfig was needed.

**Comment by martinjaeger** [comment]
> It's possible I misunderstand something, but this isn't a device. It's a software-based watchdog subsystem that someone can decide to use or not on any platform. The ways it's written though, it forces linking in the wdt driver interface, which shouldn't be required so that it an run on a system without a hardware watchdog. However, since it _can_ fall back on a hardware watchdog device, perhaps it should somehow check the DT for watchdog hardware to best determine what code needs to be compiled and linked. I'm not exactly sure how that will work b/c DT names are specific. That's why I was assuming Kconfig was needed.

It think I see what you mean and I agree. If we want to support also a pure software watchdog without any hardware watchdog (either because not being present on the hardware or because there is no driver yet), we should not `#include <drivers/watchdog.h>` here. There is `CONFIG_WATCHDOG` which we can use to determine whether a hardware watchdog driver is present. I will consider that in the next update.

@mbolivar-nordic I was wondering whether passing the device name or a `struct device *` to the `task_wdt_init()` would be better / more typical. So should `device_get_binding()` be called by the user of the `task_wdt` API even though the device could not be used in the application firmware anymore or should we call it from within the `task_wdt` itself (as currently done)?

**Comment by mbolivar-nordic** [comment]
> So should `device_get_binding()` be called by the user of the `task_wdt` API even though the device could not be used in the application firmware anymore or should we call it from within the `task_wdt` itself (as currently done)?

I don't know your use case well enough to say what the solution is, but I can say it should not assume `device_get_binding()` is being called at all, by anyone.

People have been trying to get away from the `label` property being required in Zephyr, and https://github.com/zephyrproject-rtos/zephyr/pull/30196 just got merged, which makes it possible to get a `const struct device*` directly from a `<devicetree.h>` node identifier, without any label property or string name being used.

**Comment by martinjaeger** [comment]
### Questions for API meeting

#### Storage of task watchdog channels

Currently stored in static array in `task_wdt.c` instead of a context struct. Reason: We want to be able to add a new task to the wdt from different locations in the code, so it's not convenient to pass around the context struct pointer.

#### Data passed to task watchdog callback

- Currently: Thread-ID
- Alternatives:
  - Custom string
  - `void *user_data`

### More general questions

#### Initialization of HW watchdog

The hardware watchdog can be linked to the task watchdog as an optional fallback. Its timeout should be set to the minimum task watchdog timeout. However, not all hardware watchdogs allow to change the timeout after it has been set once. Options:

1. At the beginning (in init function of task wdt). Problem: Feeding of the hardware watchdog necessary until task with shortest timeout is added
2. After shortest timeout channel has been added. Problem: Maybe too late and system already gets stuck?
3. Using extra function which is called after all tasks have been set up. Problem as above, system might have got stuck already...
4. Current solution: Add extra software channel to feed hardware watchdog if minimum timeout of all task wdt channels is larger than hardware watchdog timeout.

File: doc/reference/task_wdt/index.rst
Comment (katsuster) [inline doc/reference/task_wdt/index.rst:11]: This "fed" means "feed"?

File: subsys/task_wdt/task_wdt.c
Comment (katsuster) [inline subsys/task_wdt/task_wdt.c:120]: I think it's better to refuse "reload_period == 0" calls. It indicates invalid task_wdt, so there is no meaning for searching.

Comment (gmarull) [reply subsys/task_wdt/task_wdt.c:120]: Agree with @katsuster. Other than that, the addition of this subsys LGTM.

File: doc/reference/task_wdt/index.rst
Comment (martinjaeger) [reply doc/reference/task_wdt/index.rst:11]: "fed" is meant an alternative term for "restarted", so I used the past participle form as well. I think it is correct.

File: subsys/task_wdt/task_wdt.c
Comment (martinjaeger) [reply subsys/task_wdt/task_wdt.c:124]: I agree aswell. Fixed.

**Review by gmarull — APPROVED** [review]
I have not much experience in this area, but overall current approach LGTM and is comparable to other mentioned solutions.

File: doc/reference/task_wdt/index.rst
Comment (katsuster) [reply doc/reference/task_wdt/index.rst:11]: Ah, I see. Thanks.

**Review by katsuster — APPROVED** [review]

**Comment by martinjaeger** [comment]
Thanks for the reviews so far. Added DNM label to give a bit more time for feedback from more people.

File: doc/reference/task_wdt/index.rst
Comment (henrikbrixandersen) [inline doc/reference/task_wdt/index.rst:30]: Is this a hard requirement on the API? It seems like an odd limitation compared to many other Zephyr components.

Take for instance a use-case of having two hardware watchdogs; one for resetting the MCU itself and one for keeping an external device out of reset (e.g. an NXP Kinetis External Watchdog Monitor (EWM)). The latter is useful in safety focused applications such as motor control. It would be nice to be able to support both of these watchdogs with a "multiplexer" like the task_wdt.

File: include/task_wdt/task_wdt.h
Comment (henrikbrixandersen) [inline include/task_wdt/task_wdt.h:51]: Should this take a hardware watchdog channel number as well?

File: samples/subsys/task_wdt/prj.conf
Comment (henrikbrixandersen) [inline samples/subsys/task_wdt/prj.conf:5]: Is this option needed? It is not widely supported...

File: samples/subsys/task_wdt/sample.yaml
Comment (henrikbrixandersen) [inline samples/subsys/task_wdt/sample.yaml:18]: Is this depends line correct?

File: doc/reference/task_wdt/index.rst
Comment (nslowell) [reply doc/reference/task_wdt/index.rst:30]: I can appreciate the desire, but I think this API's scope is all about the CPU reset.  And remember that its intention is to function with or without an actual hardware CPU watchdog.  Supporting a job-specific watchdog just doesn't really seem to be the intent of this API.

Comment (martinjaeger) [reply doc/reference/task_wdt/index.rst:30]: I consider the task watchdog as something similar to a kernel service. Like the kernel timer, it hides the used hardware (system clock) in the background and has a global context, so that a new channel can be added anywhere in the application. In contrast to actual devices, we can't use `device_get_binding` in each thread to get a specific instance, that's why I prefer the global scope.

For your use case I'm wondering if it would not be sufficient to use the existing watchdog API? I guess that the EWM should mainly supervise the control loop and could still be used in parallel.

Also other RTOSes targeting safety-critical applications ([VxWorks](https://www.ee.ryerson.ca/~courses/ee8205/Data-Sheets/Tornado-VxWorks/vxworks/ref/wdLib.html), [Erika Enterprise](http://erika.tuxfamily.org/wiki/index.php?title=AUTOSAR-like_WDG_Driver)) seem to support only a single (global) task watchdog. This API does already provide more features than most other implementations, so I'd rather try to not make it too complex.

File: include/task_wdt/task_wdt.h
Comment (martinjaeger) [reply include/task_wdt/task_wdt.h:51]: Currently, `wdt_install_timeout` (for the hardware watchdog) is called inside this function, so that's also where the channel number is assigned. The application should not do anything with the hardware watchdog (except for assigning the one that it wants to use).

File: samples/subsys/task_wdt/prj.conf
Comment (martinjaeger) [reply samples/subsys/task_wdt/prj.conf:5]: According to the discussion in https://github.com/zephyrproject-rtos/zephyr/issues/23282 it's still not finally decided how startup of hardware watchdogs should be handled. In case of STM, the watchdog would be enabled at boot time if this is not set and the example would fail. So I think we have to keep this here until the hardware drivers behave consistently (e.g. keeping manufacturer defaults). Also if the watchdog timeout can be set only once, it should not be set at boot time, but within the task watchdog.

File: samples/subsys/task_wdt/sample.yaml
Comment (martinjaeger) [reply samples/subsys/task_wdt/sample.yaml:18]: Good spot. It should actually be possible to run this sample with any board (even if it doesn't have a hardware watchdog). So I'll remove the dependency.

**Comment by martinjaeger** [comment]
The last force-pushes contained some small tweaks:

- Hardware watchdog delay is now configurable through `CONFIG_TASK_WDT_HW_FALLBACK_DELAY`. I realized during testing in our application that the previously hard-coded 1 ms did not provide enough safety margin because of deviations between low-speed internal oscillator (used as clock source for the hardware watchdog in STM32) and the system clock.
- Addressed comment of @henrikbrixandersen regarding test dependencies
- Fixed sample output to be properly recognized by twister
- Added this API as "Experimental" to the API overview in the documentation

If there are no objections, I would appreciate to get this merged for 2.5 so that more people can test it in real-world applications before the LTS release.

**Comment by carlescufi** [comment]
@henrikbrixandersen @pabigot is this ready to be merged after 2.5 is released?

File: subsys/task_wdt/task_wdt.c
Comment (chrisc11) [inline subsys/task_wdt/task_wdt.c:115]: Would be neat to leverage this infrastructure for tasks created internally by zephyr as well. (i.e sysworkq, idle, shell task) but I don't see any easy way to do that. I guess in a future PR Kconfig options could be added to those task creation blocks to enable task watchdogs or do you see a better way @martinjaeger ?

**Review by chrisc11 — COMMENTED** [review]
Cool PR @martinjaeger! Left a few thoughts/ideas for future work, curious to hear what you think!

File: subsys/task_wdt/task_wdt.c
Comment (martinjaeger) [reply subsys/task_wdt/task_wdt.c:67]: As far as I understand, the `k_timer` expiry functions are called from a system clock ISR context, which is already highest priority.
Using a peripheral timer could still be a nice option to reduce computing effort. However, I'm not sure how well timer peripherals are supported for each platform in Zephyr. The use of `k_timer` makes it very generic.

File: subsys/task_wdt/task_wdt.c
Comment (martinjaeger) [reply subsys/task_wdt/task_wdt.c:87]: Yeah, good idea (and your article is also a really good read, btw). The MCU would still reset after the `z_except_reason` call, correct? If so, I could just replace the `sys_reboot` with `z_except_reason`. Wondering what others think about the best idea to reset the MCU.

File: subsys/task_wdt/task_wdt.c
Comment (martinjaeger) [reply subsys/task_wdt/task_wdt.c:115]: Interesting question. Adding task watchdogs by default is certainly not an option, as it does add some overhead and not everyone might want it. So we'd probably need a dedicated Kconfig option per internal subsystem to enable/disable the `task_wdt` supervision.

I was also thinking about whether the `task_wdt` should better become a kernel service, so it would be more generally available. And would it then make sense to integrate the hardware watchdog fallback directly into the scheduler instead of using it to monitor indirectly if `k_timer` callbacks are still happening?

File: subsys/task_wdt/task_wdt.c
Comment (chrisc11) [reply subsys/task_wdt/task_wdt.c:67]: Yeah, I like the use of k_timer by default for it's portability too. I could see being able to override with a peripheral timer when available  which would allow one to even catch hangs in timer callbacks or ISRs at a priority if configured correctly.

**Comment by carlescufi** [comment]
@henrikbrixandersen @pabigot is this ready to be merged now?

File: samples/subsys/task_wdt/sample.yaml
Comment (martinjaeger) [reply samples/subsys/task_wdt/sample.yaml]: FYI: I finally replaced this line with
```
    depends_on: watchdog
    integration_platforms:
      - nucleo_g474re
```
To limit the number of CI runs to only one and not compile it for almost every board.

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen @pabigot is this ready to be merged now?

I have tried to dismiss my review, but I am still mentioned in the list. I have previously raised my concerns with the limitations of the current implementation, but I will not object to it going in.

**Comment by pabigot** [comment]
> > @henrikbrixandersen @pabigot is this ready to be merged now?
> I have previously raised my concerns with the limitations of the current implementation, but I will not object to it going in.

That is my position as well.

**Comment by carlescufi** [comment]
Thanks @pabigot and @henrikbrixandersen. I am making a call here and merging this, subject to future improvement.

**Comment by martinjaeger** [comment]
> That is my position as well.

@pabigot I think that I have addressed your previous concerns. The software watchdog doesn't require a timer firing constantly at 1 ms interval anymore. Instead every call to the feed function will only update the timeout of a kernel timer that will never actually fire under normal operation. Also a hardware watchdog that can be configured only once is supported, as the minimum task wdt timeout (equal to hardware watchdog fallback timeout) is defined via Kconfig and not at runtime.

Does anyone have an opinion regarding the reset method, as brought up by @chrisc11? Should I replace the call to `sys_reboot` with `z_except_reason`?

**Comment by carlescufi** [comment]
> > That is my position as well.
> 
> @pabigot I think that I have addressed your previous concerns. The software watchdog doesn't require a timer firing constantly at 1 ms interval anymore. Instead every call to the feed function will only update the timeout of a kernel timer that will never actually fire under normal operation. Also a hardware watchdog that can be configured only once is supported, as the minimum task wdt timeout (equal to hardware watchdog fallback timeout) is defined via Kconfig and not at runtime.
> 
> Does anyone have an opinion regarding the reset method, as brought up by @chrisc11? Should I replace the call to `sys_reboot` with `z_except_reason`?

I did not see this before merging it. Let us continue the discussion here and submit a follow-up PR if applicable.

**Comment by martinjaeger** [comment]
> I did not see this before merging it. Let us continue the discussion here and submit a follow-up PR if applicable.

Yeah, I commented literally 1 second too late... Anyway, thanks for merging. I'm happy to create a follow-up PR if there are any further concerns :)

**Comment by nslowell** [comment]
We forgot to request tests which are good for preventing breakage and basic examples.

**Pull request merged**
